flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorAlignmentTest.java

```diff
@@ -131,6 +137,70 @@ void testWatermarkAlignmentWithTwoGroups() throws Exception {
         }
     }
 
+    /**
+     * When JobManager failover and auto recover job, SourceCoordinator will reset twice: 1. Create
+     * JobMaster --> Create Scheduler --> Create DefaultExecutionGraph --> Init
+     * SourceCoordinator(but will not start it) 2. JobMaster call
+     * restoreLatestCheckpointedStateInternal, which will call {@link
+     * org.apache.flink.runtime.operators.coordination.RecreateOnResetOperatorCoordinator#resetToCheckpoint(long,byte[])}
+     * and reset SourceCoordinator. Because the first SourceCoordinator is not be started, so the
+     * period task can't be stopped.
+     */
+    @Test
+    void testAnnounceCombinedWatermarkWithoutStart() throws Exception {
+        long maxDrift = 1000L;
+        WatermarkAlignmentParams params =
+                new WatermarkAlignmentParams(maxDrift, "group1", maxDrift);
+
+        final Source<Integer, MockSourceSplit, Set<MockSourceSplit>> mockSource =
+                createMockSource();
+
+        // First to init a SourceCoordinator to simulate JobMaster init SourceCoordinator
+        AtomicInteger counter1 = new AtomicInteger(0);
+        sourceCoordinator =
+                new SourceCoordinator<MockSourceSplit, Set<MockSourceSplit>>(
+                        OPERATOR_NAME,
+                        mockSource,
+                        getNewSourceCoordinatorContext(),
+                        new CoordinatorStoreImpl(),
+                        params,
+                        null) {
+                    @Override
+                    void announceCombinedWatermark() {
+                        counter1.incrementAndGet();
+                    }
+                };
+
+        // Second we call SourceCoordinator::close and re-init SourceCoordinator to simulate
+        // RecreateOnResetOperatorCoordinator::resetToCheckpoint
+        sourceCoordinator.close();
+        AtomicInteger counter2 = new AtomicInteger(0);
+        CountDownLatch latch = new CountDownLatch(2);
+        sourceCoordinator =
+                new SourceCoordinator<MockSourceSplit, Set<MockSourceSplit>>(
+                        OPERATOR_NAME,
+                        mockSource,
+                        getNewSourceCoordinatorContext(),
+                        new CoordinatorStoreImpl(),
+                        params,
+                        null) {
+                    @Override
+                    void announceCombinedWatermark() {
+                        counter2.incrementAndGet();
+                        latch.countDown();
+                    }
+                };
+
+        sourceCoordinator.start();
+        setReaderTaskReady(sourceCoordinator, 0, 0);
+
+        latch.await();
+        assertThat(counter1.get()).isZero();
+        assertThat(counter2.get()).isGreaterThanOrEqualTo(2);
```

Comment:
The `counter2` isn't necessary. The `counter2.get()` must be GreaterThanOrEqualTo 2 after `latch.await()` is finished.

Reply:
updated~ @1996fanrui